File: NC/level-order-traversal-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        vector<vector<int>> ans;

        queue<pair<TreeNode *, int>> nodeLevelList;

        nodeLevelList.push({root, 1});

        while (nodeLevelList.size() > 0)
        {
            TreeNode *currentNode = nodeLevelList.front().first;
            int currentLevel = nodeLevelList.front().second;
            nodeLevelList.pop();

            if (!currentNode)
            {
                continue;
            }

            if (ans.size() < currentLevel)
            {
                ans.push_back({});
            }

            ans[currentLevel - 1].push_back(currentNode->val);

            nodeLevelList.push({currentNode->left, currentLevel + 1});
            nodeLevelList.push({currentNode->right, currentLevel + 1});
        }

        return ans;
    }
};

File: NC/java/LevelOrderTraversalOfBinaryTree.java
package NC.java;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

import javax.swing.tree.TreeNode;

public class LevelOrderTraversalOfBinaryTree {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        Queue<TreeNode> q = new LinkedList<>();
        if (root == null) {
            return ans;
        }
        q.add(root);
        q.add(null);
        ans.add(new ArrayList<>());
        while (q.size() > 0) {
            TreeNode curr = q.poll();
            if (curr == null) {
                if (q.size() > 0) {
                    ans.add(new ArrayList<>());
                    q.add(null);
                }
            } else {
                ans.get(ans.size() - 1).add(curr.val);
                if (curr.left != null) {
                    q.add(curr.left);
                }
                if (curr.right != null) {
                    q.add(curr.right);
                }
            }
        }
        return ans;
    }
}
